### PR TITLE
Add serialization for std::set (as there is for std::vector and std::map)

### PR DESCRIPTION
**Pull request by gentryx**

### CODE REVIEW

**Comment by hkaiser** [comment]
Inspect is complaining: https://circle-artifacts.com/gh/STEllAR-GROUP/hpx/974/artifacts/0/tmp/circle-artifacts.8vvqUYx/hpx_inspect_report.html

**Comment by gentryx** [comment]
Oops. My bad, my name. Changed to "ae"

File: hpx/runtime/serialization/set.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    template <typename T, typename Allocator>
+    void serialize(input_archive & ar, std::set<T, Allocator> & set, unsigned)
+    {
+        set.clear();
+        std::size_t size;
+        ar >> size;
+        for (std::size_t i = 0; i < size; ++i) {
+            T t;
+            ar >> t;
+            set.insert(set.end(), t);
```
Comment (sithhell) [inline hpx/runtime/serialization/set.hpp]: You should make use of move semantics here.

File: hpx/runtime/serialization/set.hpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#define HPX_SERIALIZATION_SET_HPP
+
+#include <hpx/runtime/serialization/serialize.hpp>
+
+#include <set>
+
+namespace hpx { namespace serialization
+{
+    template <typename T, typename Allocator>
+    void serialize(input_archive & ar, std::set<T, Allocator> & set, unsigned)
```
Comment (sithhell) [inline hpx/runtime/serialization/set.hpp]: The `Allocator` is not the second template argument, `Compare` is.

File: hpx/runtime/serialization/set.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        ar >> size;
+        for (std::size_t i = 0; i < size; ++i) {
+            T t;
+            ar >> t;
+            set.insert(set.end(), t);
+        }
+    }
+
+    template <typename T, typename Allocator>
+    void serialize(output_archive & ar, std::set<T, Allocator> & set, unsigned)
```
Comment (sithhell) [inline hpx/runtime/serialization/set.hpp]: Same as above...

File: hpx/runtime/serialization/set.hpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        ar << set.size(); //-V128
```
Comment (sithhell) [inline hpx/runtime/serialization/set.hpp]: please don't use `std::size_t`  here, this is platform dependant and might break in heterogeneous environments. a fied size type like `boost::uint64_t` would be more appropriate.

Comment (gentryx) [inline hpx/runtime/serialization/set.hpp]: ACK. should I fix this for vector and map as well?

Comment (sithhell) [inline hpx/runtime/serialization/set.hpp]: Am 15.07.2015 3:19 nachm. schrieb "Andreas Schäfer" <
notifications@github.com>:
>
> In hpx/runtime/serialization/set.hpp:
>
> > +    void serialize(input_archive & ar, std::set<T, Allocator> & set,
unsigned)
> > +    {
> > +        set.clear();
> > +        std::size_t size;
> > +        ar >> size;
> > +        for (std::size_t i = 0; i < size; ++i) {
> > +            T t;
> > +            ar >> t;
> > +            set.insert(set.end(), t);
> > +        }
> > +    }
> > +
> > +    template <typename T, typename Allocator>
> > +    void serialize(output_archive & ar, std::set<T, Allocator> & set,
unsigned)
> > +    {
> > +        ar << set.size(); //-V128
>
> ACK. should I fix this for vector and map as well?

That would be splendid!

Comment (hkaiser) [inline hpx/runtime/serialization/set.hpp]: @gentryx Please do that on a separate PR, though

**Comment by gentryx** [comment]
All comments should be addressed by now. Once this is merged I'll follow up with a separate PR to update map.hpp and vector.hpp.

**Comment by AntonBikineev** [comment]
@sithhell using size_t shouldn't be a problem and user doesn't need to be aware of it, because it's explicitly casted to int32_t/uint32_t in archives' code.

File: hpx/runtime/serialization/set.hpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        }
+    }
+
+    template <class T, class Compare, class Allocator>
+    void serialize(output_archive & ar, std::set<T, Compare, Allocator> & set, unsigned)
+    {
+        boost::uint64_t size = set.size();
+        ar << size;
+        if(set.empty()) return;
+        for (typename std::set<T, Allocator>::iterator i = set.begin(); i != set.end(); ++i) {
```
Comment (hkaiser) [inline hpx/runtime/serialization/set.hpp]: This does not look right, Allocator is the last template argument. Why not write this as 

    for (T const& elem: set) ar << elem;

to avoid problems?

**Comment by hkaiser** [comment]
@AntonBikineev I hope it does not get casted to a 32bit but a 64bit value inside the archive..

File: hpx/runtime/serialization/set.hpp (hunk 5 above)
Comment (gentryx) [inline hpx/runtime/serialization/set.hpp]: ACK

**Comment by AntonBikineev** [comment]
@hkaiser any integer is always casted to 64 bit value. BTW, that behaviour may lead to narrowing using something like boost::uint128_type.

**Comment by gentryx** [comment]
@AntonBikineev So if I have 1000 int they'll take up 8k, not 4k? Is that what we want?

**Comment by sithhell** [comment]
Am 15.07.2015 8:05 nachm. schrieb "Anton Bikineev" <notifications@github.com
>:
>
> @sithhell using size_t shouldn't be a problem and user doesn't need to be
aware of it, because it's explicitly casted to int32_t/uint32_t in
archives' code.

That sounds awfully wrong, I hope that's not the case. We've been bitten by
using size_t inside of serialization before. There shouldn't be any code
that's used for serialization that relies on std::size_t.

**Comment by AntonBikineev** [comment]
@sithell my understanding is following. when you use
```
size_t size;
ar << size
```
it calls
```
template <class T>
boost::enable_if<boost::is_integral<T>::value || boost::is_unsigned<T>::value> save(T t) // or so
{
    save(static_cast<boost::uint64_t>(t));
}

**Comment by sithhell** [comment]
On 07/15/2015 08:57 PM, Anton Bikineev wrote:
> @sithell my understanding is following. when you use
>
> |size_t size; ar << size |
>
> it calls
>
> |template <class T> boost::enable_if<boost::is_integral<T>::value ||
> boost::is_unsigned<T>::value> save(T t) // or so {
> save(static_cast<boost::uint64_t>(t)); } |

You are right ... my bas. Some issues we've had in the past with 
serializing std::size_t was when values like std::size_t(-1) might 
appear, which is dangerous when getting sent from a platform where 
sizeof(std::size_t(-1)) is 4 bytes to a platform where it is 8 bytes.

**Comment by hkaiser** [comment]
> BTW, that behaviour may lead to narrowing when serializing something like boost::uint128_type. 

@AntonBikineev Could you verify what happens for uint128_t, please?

**Comment by hkaiser** [comment]
The PR LGTM now.

**Comment by AntonBikineev** [comment]
@hkaiser yeah, just reproduced this with the help of debugger. A half of most significant bits are cut off. What do you think would be more appropriate here? Having overload for this or just leave it as is to not mess the code?

**Comment by hkaiser** [comment]
Could you please add an appropriate overload for platforms supporting that type?

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://nissanfethiye.com/garden.php?xp06>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://wkthai.com/even.php?4>

 

mario.mulansky@gmx.net

File: hpx/runtime/serialization/set.hpp (hunk 4 above)
Comment (mariomulansky) [inline hpx/runtime/serialization/set.hpp]: Hello!

 

New message, please read <http://grafikornekleri.com/move.php?ipc>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://learnsex.net/until.php?qv7>

 

mario.mulansky@gmx.net

File: hpx/runtime/serialization/set.hpp (hunk 3 above)
Comment (mariomulansky) [inline hpx/runtime/serialization/set.hpp]: Hello!

 

New message, please read <http://novatherm.it/age.php?dfdt>

 

mario.mulansky@gmx.net

File: hpx/runtime/serialization/set.hpp (hunk 4 above)
Comment (mariomulansky) [inline hpx/runtime/serialization/set.hpp]: Hello!

 

New message, please read <http://sokhandani.com/quick.php?5von>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://sharkbaratl.com/floor.php?w5mcw>

 

mario.mulansky@gmx.net

File: hpx/runtime/serialization/set.hpp (hunk 4 above)
Comment (mariomulansky) [inline hpx/runtime/serialization/set.hpp]: Hello!

 

New message, please read <http://www.taimeipotato.com/coming.php?3aefl>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://genovayacht.com/paper.php?k>

 

mario.mulansky@gmx.net

File: hpx/runtime/serialization/set.hpp (hunk 5 above)
Comment (mariomulansky) [inline hpx/runtime/serialization/set.hpp]: Hello!

 

New message, please read <http://kiddyland.com.sg/directly.php?u6jm4>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://cesmetekneturu.com/rich.php?xis>

 

mario.mulansky@gmx.net

File: hpx/runtime/serialization/set.hpp (hunk 4 above)
Comment (mariomulansky) [inline hpx/runtime/serialization/set.hpp]: Hello!

 

New message, please read <http://fatuzhushou.cn/virtual.php?m3tbz>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://capital-finance093.tk/conviction.php?zi26>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://montanacash.hu/team.php?8l>

 

mario.mulansky@gmx.net

File: hpx/runtime/serialization/set.hpp (hunk 5 above)
Comment (mariomulansky) [inline hpx/runtime/serialization/set.hpp]: Hello!

 

New message, please read <http://polatlicocukdiyari.com/increasing.php?0ooah>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

New message, please read <http://kcsr.ba/deny.php?43s>

 

mario.mulansky@gmx.net

**Comment by mariomulansky** [comment]
Hello!

 

You have a new message, please read <http://schoenenhype.nl/building.php?8kh6>

 

mario.mulansky@gmx.net
